### PR TITLE
Add an option to specify components to DependencyUpdateCheck

### PR DESCRIPTION
**Pull request by querti**
For now, the full functionality is not implemented, the controller only
displays a message and exits when encountering DependencyUpdateCheck
with specified components. The controller logic will be implemented as a
part of the next Jira.

Components are specified in a nested structure of workspace name -
application name - component name. It is possible to omit application
name or component name, in which case all applications/components in
a given workspace/application will be processed.

Please note that some changes are autogenerated by kubebuilder.

Refers to CLOUDDST-24331

### CODE REVIEW

File: api/v1alpha1/dependencyupdatecheck_types.go [hunk 1]
```diff
@@ -28,8 +28,10 @@ type DependencyUpdateCheckSpec struct {
 	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
 
-	// Foo is an example field of DependencyUpdateCheck. Edit dependencyupdatecheck_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	// Specifies which components to run Renovate on. The expected format is "workspace/component".
+	// If omitted, Renovate will be run on all components
+	// +optional
+	Components []string `json:"components,omitempty"`
```
Comment (qixiang) [inline api/v1alpha1/dependencyupdatecheck_types.go:34]: I'm thinking about how to use a more flexible field for specifying components, and provide the functionality for validating the CR while creating it. What do you think about defining a schema in format like this:
```
workspaces:
  workspace1:
    applications:
      app1:
        components:
        - component1
        - component2
  workspace2:
    applications:
      app1:
        components:
        - componentA
        - componentB
      app2:
        components:
        - component1
        - component2
```
Note: I believe it's possible to create components with same name in different applications, and application with same name in different workspaces.

Comment (querti) [reply api/v1alpha1/dependencyupdatecheck_types.go]: I changed the input to a nested structure and added some kubebuilder validations

**Review by qixiang — COMMENTED** [review]
Just minor suggestion to the field names:
`applicationName` -> `application`
`workspaceName` -> `workspace`
this change could make it more user friendly while creating new CRs by manual, we don't support that yet, but it's possible in the future. 

EDIT: not sure why my pending comments were lost while submitting the review. I've re-added some suggestion comments, but didn't cover all the lines need to be changed.

File: api/v1alpha1/dependencyupdatecheck_types.go [hunk 2]
```diff
@@ -20,6 +20,36 @@ import (
...
+// +kubebuilder:validation:Pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
+// +kubebuilder:validation:MaxLength=63
+type Component string
+
+type Application struct {
+	// +kubebuilder:validation:Pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
+	// Specifies the name of the application for which to run Mintmaker.
+	// Required.
+	// +required
+	ApplicationName string `json:"applicationName"`
```
Comment (qixiang) [inline api/v1alpha1/dependencyupdatecheck_types.go:32]:
```suggestion
type ApplicationSpec struct {
	// +kubebuilder:validation:Pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
	// Specifies the name of the application for which to run Mintmaker.
	// Required.
	// +required
	Application string `json:"application"`
```

File: api/v1alpha1/dependencyupdatecheck_types.go [hunk 3]
```diff
@@ -20,6 +20,36 @@ import (
...
+	// +optional
+	Components []Component `json:"components,omitempty"`
+}
+
+type Workspace struct {
+	// +kubebuilder:validation:Pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
+	//Specifies the name of the workspace for which to run Mintmaker.
+	// Required.
+	// +required
+	WorkspaceName string `json:"workspaceName"`
```
Comment (qixiang) [inline api/v1alpha1/dependencyupdatecheck_types.go:45]:
```suggestion
type WorkspaceSpec struct {
	// +kubebuilder:validation:Pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
	//Specifies the name of the workspace for which to run Mintmaker.
	// Required.
	// +required
	Workspace string `json:"workspace"`
```

File: config/samples/appstudio_v1alpha1_dependencyupdatecheck.yaml [hunk 4]
```diff
@@ -7,6 +7,14 @@ metadata:
     app.kubernetes.io/part-of: mintmaker
     app.kubernetes.io/managed-by: kustomize
     app.kubernetes.io/created-by: mintmaker
-  name: dependencyupdatecheck-sample
+  generateName: dependencyupdatecheck-
 spec:
-  # TODO(user): Add fields here
+  workspaces:
+  - workspaceName: "workspace1-tenant"
```
Comment (qixiang) [inline config/samples/appstudio_v1alpha1_dependencyupdatecheck.yaml:13]:
```suggestion
  - workspace: "workspace1"
```
`{workspace}-tenant` is the namespace name in Konflux.

**Comment by querti** [comment]
> Just minor suggestion to the field names: `applicationName` -> `application` `workspaceName` -> `workspace` this change could make it more user friendly while creating new CRs by manual, we don't support that yet, but it's possible in the future.
> 
> EDIT: not sure why my pending comments were lost while submitting the review. I've re-added some suggestion comments, but didn't cover all the lines need to be changed.

I renamed the parameters.

File: internal/controller/dependencyupdatecheck_controller.go [hunk 5]
```diff
@@ -102,6 +102,12 @@ func (r *DependencyUpdateCheckReconciler) Reconcile(ctx context.Context, req ctr
 		return ctrl.Result{}, nil
 	}
 
+	if len(dependencyupdatecheck.Spec.Workspaces) > 0 {
+		log.Info(fmt.Sprintf("Following components are specified: %v", dependencyupdatecheck.Spec.Workspaces))
+		log.Error(err, "The functionality to run Renovate only on the specified components is not yet implemented")
```
Comment (qixiang) [inline internal/controller/dependencyupdatecheck_controller.go:107]:
```suggestion
		log.Error("The functionality to run Renovate only on the specified components is not yet implemented")
```

**Review by qixiang — COMMENTED** [review]
I created a CR with workspaces specified, then ran the controller, however seems `dependencyupdatecheck.Spec.Workspaces` is always empty, I haven't figured out what is missing, could you give it a try?

Here is my steps of testing this:

1. Boot up a k8s cluster with minikube. BTW, I tried kind as well, however seems controller doesn't receive the event of CR creation, minikube works well.

2. Create a new namespace `mintmaker`. The controller only react upon `DependencyUpdateCheck` CRs created in `mintmaker` namespace.

3. [optinoal]: Install the application-api CRDs.
    ```
    $ kubectl apply -k https://github.com/konflux-ci/application-api/config/crd
    ```
    This provides the CRD of `application` and `component` which are required if we want to test with these CRDs.
4. Install MintMaker `DependencyUpdateCheck` CR:
    ```
    $ kubectl apply -k config/crd/
    ```
    this is sufficient for this spec change, but if you want to test more, you need to install other manifests too:
    ```
    $ kubectl apply -k config/default/
    $ kubectl apply -k config/renovate/
    ```
5. Run the controller.
    ```
    $ go run cmd/main.go
    ````
    or run with debugger:
    ```
    $ dlv debug cmd/main.go
    ```
 6. Create a testing CR in another tab.
     ```
     $ kubectl create -f /path/to/cr.yaml
     ```
     You will observe the log in controller running tab.

File: internal/controller/dependencyupdatecheck_controller.go (hunk 5 above)
Comment (querti) [reply internal/controller/dependencyupdatecheck_controller.go]: Error logs message must have error as the first parameter, so I changed the second log message to info. I hope that's OK.

**Comment by qixiang** [comment]
LGTM, tested with an OpenShift cluster (created by clusterbot), I can now see the workspaces in CR's spec. It was likely an issue with minikube or my testing environment that initially prevented me from viewing the correct spec.

**Review by qixiang — APPROVED** [review]

**Pull request merged**